Add user login

qptr opened the pull request:
Add sign out button

qptr committed:
Integrate google sign in into app
Add sign out button

qptr committed:
Restructure project

qptr committed:
Fix login redirect event handling

qptr committed:
Fix user logout from login and home pages

qptr committed:
Send id token to backend for verification

qptr committed:
Make response body a map

qptr committed:
Fix initialize-db test

qptr committed:
Use find-by-keys function to fetch record by google id
Make functions private
Add try catch to catch error thrown by bad id token
Respond with status 400 if id token header is missing
Return default db on user logout for now

qptr committed:
Add events tests on f/e, handler and db tests for b/e

qptr committed:
Add session namespace
Set session id cookie in login handler
Add session id check to fetch books handler
Put redis host in config
Fix handler tests
Add redis service to ci for backend tests

jysandy commented on src/clj/athenaeum/handlers/user.clj:
```diff
@@ -0,0 +1,39 @@
+  []
+  1)
+
+(defn- confirm-login
```
This can be split into: find or create the user, and create the session. Then we can do away with this `confirm-login` name, which is not very meaningful.

jysandy commented on src/clj/athenaeum/handlers/user.clj:
```diff
@@ -0,0 +1,45 @@
+                             (catch Exception _ nil))]
+      (walk/keywordize-keys (into {} (.getPayload id-token))))))
+
+(defn- get-session-id
```
This can be split into: find or create the user, and create the session.

jysandy commented on src/clj/athenaeum/session/redis.clj:
```diff
@@ -0,0 +1,33 @@
+(ns athenaeum.session.redis
```
This doesn't need to be under `session`, because all of the functions are independent of sessions.

jysandy commented on src/cljs/athenaeum_web/login_page/events.cljs:
```diff
@@ -0,0 +1,30 @@
+   (assoc db :login-state :logged-out)))
+
+(rf/reg-event-fx
+ ::send-id-token
```
I would rename this to something like `::login`.

jysandy commented on src/cljs/athenaeum_web/login_page/events.cljs:
```diff
@@ -0,0 +1,30 @@
+(rf/reg-event-db
+ ::add-id-token
+ (fn [db [_ id-token]]
+   (assoc-in db [:user :id-token] id-token)))
```
Do we need the id-token in the database?

qptr replied:
just until this story is over.

jysandy commented on src/clj/athenaeum/domain/user.clj:
```diff
@@ -0,0 +1,17 @@
+  [c user]
+  (sql/insert! c
+               :users
+               {:google_id (:sub user)
```
It's a little strange that this function expects the google ID to be under the `:sub` key. We can do away with that, and look up the google ID from the `:google-id` key or something, which seems more appropriate.

jysandy replied:
Obviously the callers will have to be changed to pass in the map with the appropriate keys.

jysandy commented on src/clj/athenaeum/domain/user.clj:
```diff
@@ -0,0 +1,17 @@
+               {:google_id (:sub user)
+                :name      (:name user)
+                :email     (:email user)
+                :photo_url (:picture user)}))
```
Similarly, why not look up the `:photo-url` key instead?

jysandy commented on src/clj/athenaeum/session/redis.clj:
```diff
@@ -0,0 +1,33 @@
+
+(defn get-value
+  [key]
+  (when (= 1 (wcar* (car/exists key)))
```
This should not be necessary. `get` will return `nil` if the key doesn't exist.

jysandy commented on src/clj/athenaeum/session/redis.clj:
```diff
@@ -0,0 +1,33 @@
+(defn set-key
+  [key value]
+  (wcar* (car/set key value)
+         (car/expire key 3600)))
```
This function should probably accept the TTL as an optional argument. Only the session-specific namespace should know about the TTL, and the TTL itself should be in config.

jysandy commented on test/clj/athenaeum/domain/user_test.clj:
```diff
@@ -0,0 +1,35 @@
+          returned-user (db/with-transaction [tx @db/datasource]
+                          (user/fetch-by-google-id tx google-id))
+          remove-id #(dissoc % :id)]
+      (is (= (remove-id user) (remove-id returned-user)))))
```
Why remove the id? The ids should be the same.

jysandy commented on test/clj/athenaeum/domain/user_test.clj:
```diff
@@ -0,0 +1,35 @@
+    (tu/clear-tables)
+    (let [google-id-1 "test-id-1"
+          _user (db/with-transaction [tx @db/datasource]
+                  (user/create tx {:sub     google-id-1
```
I wouldn't bother creating a user in this test.

jysandy commented on test/clj/athenaeum/handlers/book_test.clj:
```diff
@@ -4,21 +4,36 @@
+            [athenaeum.session.core :as session]
+            [athenaeum.session.redis :as redis]))
 
 (use-fixtures :once fixtures/load-config fixtures/set-datasource)
```
Don't we need to connect to Redis here?

jysandy commented on src/clj/athenaeum/handlers/book.clj:
```diff
@@ -1,9 +1,15 @@
-  (db/with-transaction [tx @db/datasource]
-    (response/response (books/fetch-all tx))))
+  [{:keys [cookies]}]
+  (if-let [session-id (get cookies "session-id")]
```
This checking should go into a middleware. Then it can be maintained and tested separately.

jysandy commented on test/clj/athenaeum/handlers/user_test.clj:
```diff
@@ -0,0 +1,28 @@
+  (testing "Returns status 200, adds id-token cookie to response on successful id token verification"
+    (let [req {:headers {:id-token "valid-id-token"}}
+          res (with-redefs [user/verify-id-token (constantly true)
+                            user/get-session-id (constantly "new-session-id")]
```
Instead of mocking this here, connect to Redis in the test and verify that the session key is set as expected.

jysandy reviewed (changes requested):
Added some comments. Frontend tests seem okay so far from what I can tell.

qptr committed:
Hard redirect to home page on log out

- Remove view for handling logged out state
- Fix on-click property of button
- Fix init db state to include handler

qptr committed:
Delete user session on logout

- Add request handler to delete session
- Add event handler to make get rq to /api/logout
- Add logout request handler tests
- Fix session-id cookie destructuring
- Convert session id to str from java.util.UUID object

qptr committed:
Refactor according to pr comments

- Split get-session-id into find-or-create-user and create-session
- Move session id creation to separate function
- Don't stub session creation in login handler test

qptr committed:
Refactor according to pr comments

- Put ttl of session in config file
- Rename send id token event to login
- Truncate all tables with clear-tables test utility fn
- Rename history-token effect to navigate-to

qptr committed:
Connect to redis in tests

- Add redis host and ttl to ci and test config files
- Pass in redis host to set-conn-opts correctly
- Call set-conn-opts in core and dev.repl-utils
- Add test fixture to set redis server connection options with test config
- Add test fixture to clear redis
- Add function to reset server-conn atom

qptr committed:
Changes according to pr comments

- Add function to reformat payload map
- Add check in login handler so only users with nilenso.com emails can login
- Add test for ^
- Redirect to ogin page on logout
- Use replace-token instead of set-token
- Don't remove id's in book-test

qptr committed:
Refactor request handler tests

- Add with-fixtures macro to wrap fixtures around each testing block
- Use session level functions instead of redis level ones in tests
- Rename clear-redis fixture to clear-sessions
- Add session level function to delete all redis keys

qptr committed:
Redirect to home page on successful login

- Navigate to / on login-success event
- Add login-page-navigated event to redirect to / when logged in
- Remove go home button component on login page

qptr committed:
Check user session validity every refresh to keep user logged in

- Add handler to check if session corresponding to session id in cookie exists
- Add event handler to make get rq to /api/me
- Dispatch this event before every page render
- Refactor clear-tables test fixture

scimetfoo committed:
[+Akshatha] Make wrap-keywordize-cookies-and-headers a higher order
function to accept a verification middleware

Fix tests failing since middleware was added

qptr committed:
Refactor middleware ns

- Use comp to compose the middleware
- Add middleware tests
- Set path field to session id cookie to specify its scope
- Change routes so /user paths are under /api

qptr committed:
Add /me api

- Add function to fetch user by id
- Add handler to fetch user if session id is valid
- Add event handler to fetch user and assoc in db
- Dispatch this in initialize-db

qptr committed:
Add frontend tests and test utils

- Add tests for app, login-page, home-page and page events
- Add test utils to initialize db, stub events and stub effects

qptr committed:
Call parseInt only on non-nil user ids

qptr committed:
Update README

- Instructions to run migrations and install Redis and Postgres
- Instructions to setup dev and test db's
- Split into more sections

qptr committed:
Minor refactoring

- Put fetch-books db call in try catch block
- Add wrap-content-type middleware for js warning
- Make navigate-to namespace qualified

qptr committed:
Reorder events dispatched on page refresh

- Fetch user takes on-success and on-failure callback events
- Initialize-db only resets db to default-db
- Default and home page navigation pass redirect-to-login callback if fetch-user fails
- Update home-page-navigated and initialize-db tests

qptr committed:
Load gapi.auth2 library before ui loads

- Track load status of library with :auth2-loaded field in db
- Initialize db with :auth2-loaded set to false
- As soon as it loads, specify the app's client id with auth2.init() method
- Also dispatch event to update :auth2-loaded to true
- Logout button is disabled until auth2 is loaded, as signOut() depends on auth2 library
- Retain auth2-loaded status event after logging out

qptr committed:
Minor changes

- Update reccommended postgres version in README
- Install, enable and restart redis server on staging

qptr committed:
Restructure cljs events namespaces and add more tests

- Make authentication checking a separate event
- Add routing and authentication events ns
- Rename aliases more clearly
- Put map-id-to-value fn into utils ns
- Remove add-id-token event
- Add more thorough testing blocks for event handlers

qptr committed:
Add exception handling middleware and tests

qptr committed:
Move session validation into a middleware

- Add tests for said middleware
- Update tests for the now changed handlers
- Refactor the tests

qptr committed:
Minor changes

- Refactor user handlers ns
- Wrap-require-session-id-cookie for logout so user is logged out even after session expiry without error
- Dispatch fetch-user on login success
- Update login handler test

jysandy commented on src/clj/athenaeum/handlers/user.clj:
```diff
@@ -0,0 +1,83 @@
+(defn- verify-id-token
+  [id-token-string]
+  (let [client-id (get-in @config/config [:oauth :google :client-id])
+        id-token-verifier (-> (GoogleNetHttpTransport/newTrustedTransport)
```
For better performance, instead of building the verifier in each function call, it should be built only once and used multiple times.

jysandy commented on src/clj/athenaeum/handlers/user.clj:
```diff
@@ -0,0 +1,83 @@
+       (walk/keywordize-keys)
+       (reformat-payload)))
+
+(defn- find-or-create-user
```
This can go into `domain.user`.

jysandy commented on src/clj/athenaeum/handlers/user.clj:
```diff
@@ -0,0 +1,83 @@
+       (reformat-payload)))
+
+(defn- find-or-create-user
+  [payload]
```
Try to find a better name for this after it's moved to `domain.user`.

jysandy commented on src/clj/athenaeum/middleware.clj:
```diff
@@ -0,0 +1,53 @@
+          (response/status 401)))))
+
+(def wrap-require-id-token-header
+  (comp wrap-keywordize-cookies-and-headers
```
Instead of composing this middleware like this, you can just include it separately in your middleware stack.

jysandy commented on src/clj/athenaeum/middleware.clj:
```diff
@@ -0,0 +1,53 @@
+      (handler request)
+      (response/bad-request {:message "id token header missing"}))))
+
+(defn wrap-require-session-id-cookie*
```
Why is this middleware necessary? `wrap-require-session` should suffice.

qptr replied:
i thought for logout clicking on logout should log you out even if your session expired after an hour instead of returning an error? but i guess i could later add a check to automatically logout if session expires like you said.

qptr replied:
this was for the logout handler so it would logout the user without error even if the session had expired already. but i guess i could later add the check to logout automatically when the session expires like you said.

jysandy commented on src/clj/athenaeum/middleware.clj:
```diff
@@ -0,0 +1,53 @@
+                 (update :cookies walk/keywordize-keys)
+                 (update :headers walk/keywordize-keys)))))
+
+(defn wrap-require-id-token-header*
```
It doesn't make sense to have this as a middleware, because only one API requires it.

jysandy commented on src/clj/athenaeum/middleware.clj:
```diff
@@ -0,0 +1,53 @@
+(defn wrap-require-id-token-header*
+  [handler]
+  (fn [{:keys [headers] :as request}]
+    (if (:id-token headers)
```
Auth tokens are typically provided under the `Authorization` header, with the format of the value as `Bearer <token>`.

jysandy commented on src/clj/athenaeum/session.clj:
```diff
@@ -0,0 +1,33 @@
+(defn create
+  "returns session id"
+  [user-id]
+  (let [session user-id
```
This binding isn't needed.

jysandy commented on src/cljs/athenaeum_web/app/events/authentication.cljs:
```diff
@@ -0,0 +1,17 @@
+   {:fx [[::effects/navigate-to (routes/path-for :login-page)]]}))
+
+(rf/reg-event-fx
+ ::authentication-check
```
Should this be an event or a function? Think about it.

jysandy commented on src/cljs/athenaeum_web/app/events/core.cljs:
```diff
@@ -0,0 +1,38 @@
+ (fn [{:keys [db]} [_ event user]]
+   {:db       (-> db
+                  (assoc :login-state :logged-in)
+                  (assoc :user (:user user)))
```
`(:user user)` looks strange, you can name the response `response` or destructure `user` instead.

jysandy commented on test/clj/athenaeum/handlers/user_test.clj:
```diff
@@ -0,0 +1,72 @@
+                                                            :picture   "picture"})
+                              session/new-id (constantly "new-session-id")]
+                  (user/login req))]
+        (is (= 401 (:status res)))
```
This should return a `400`, not a `401`. `401` means that the failure was because the client needs to authenticate to access the resource, which doesn't make sense for a login API. `400` means that there was bad or invalid data in the request.

jysandy commented on src/clj/athenaeum/handlers/user.clj:
```diff
@@ -0,0 +1,83 @@
+                       :session-id
+                       session-id
+                       {:same-site :strict
+                        :max-age   3600
```
This value should come from config, and it should be the same as the session TTL in Redis.

jysandy commented on test/cljs/athenaeum_web/events/login_page_test.cljs:
```diff
@@ -0,0 +1,43 @@
+       (rf/dispatch [::login-page-events/login-page-navigated])
+       (is (= ["/"] @navigate-to-params)))))
+
+  (testing "When login page is navigated and user is logged in, nothing happens"
```
I wasn't able to follow this test. Can you explain what is being tested here?
Also the docstring says `logged in`, but the test is setting the state to `:logged-out`.

qptr replied:
that's supposed to say logged *out. should i just remove it. i added it because the test only said what happens when the user is logged in, but it's unnecessary otherwise.

qptr committed:
Changes according to pr comments

qptr committed:
Changes according to pr comments

qptr committed:
Put all routing logic in routes ns

qptr committed:
Call set-id-token-verifier in main fn

jysandy commented on config/config.ci.edn:
```diff
@@ -4,4 +4,5 @@
                :password "athenaeum_ci_pwd"
-               :host     "localhost"}}
+               :host     "localhost"}
+ :oauth       {:google {:client-id "391229463438-a0vus34lglv9tsggpbmq2l1ekfb2lc3s.apps.googleusercontent.com"}}}
```
Is this client ID necessary? Maybe just change it to a random string like `"dummy"`?

qptr replied:
it's the old invalid client id

jysandy commented on src/clj/athenaeum/handlers/user.clj:
```diff
@@ -35,49 +44,47 @@
-      (-> (response/response {:message "invalid domain"})
-          (response/status 401)))
-    (response/bad-request {:message "id token verification failed"})))
+  (if-let [id-token (-> headers
```
You should use `some->` here, otherwise `str/split` might throw in case of a `nil`.

jysandy commented on src/clj/athenaeum/handlers/user.clj:
```diff
@@ -35,49 +44,47 @@
+                                         (user/find-by-google-id-or-create tx user))
+                                       :id
+                                       session/create)))
+        (-> (response/response {:message "invalid domain"})
```
Nit: you can use `response/bad-request` here.

jysandy commented on src/cljs/athenaeum_web/login_page/events.cljs:
```diff
@@ -15,14 +13,14 @@
    {:http-xhrio {:method          :post
                  :uri             "/api/user/login"
-                 :headers         {:id-token id-token}
+                 :headers         {:Authorization (str "Bearer " id-token)}                          ;{:id-token id-token}
```
Looks like a stray comment.
```suggestion
                 :headers         {:Authorization (str "Bearer " id-token)}
```

qptr committed:
Changes according to pr comments
